lib/circleci/token_validator_test.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package circleci
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// fakeIDP pretends to be a circle CI org OIDC provider, e.g:
+// https://oidc.circleci.com/org/00f45056-5918-4171-9222-b538c76ed3f8/.well-known/openid-configuration
+type fakeIDP struct {
+	signer jose.Signer
+	server *httptest.Server
+}
+
+func (f *fakeIDP) close() {
+	f.server.Close()
+}
+
+func (f *fakeIDP) issueToken(
+	t *testing.T,
+	organizationID string,
+	projectID string,
+	contextIDs []string,
+	issuedAt time.Time,
+	expiry time.Time,
+) string {
+	stdClaims := jwt.Claims{
+		Issuer:   fmt.Sprintf(f.issuerURLTemplate(), organizationID),
+		Subject:  fmt.Sprintf("org/%s/project/%s/user/USER_ID", organizationID, projectID),
+		Audience: jwt.Audience{organizationID},
+		IssuedAt: jwt.NewNumericDate(issuedAt),
+		Expiry:   jwt.NewNumericDate(expiry),
+	}
+	customClaims := map[string]interface{}{
+		"oidc.circleci.com/project-id":  projectID,
+		"oidc.circleci.com/context-ids": contextIDs,
+	}
+	token, err := jwt.Signed(f.signer).
+		Claims(stdClaims).
+		Claims(customClaims).
+		CompactSerialize()
+	require.NoError(t, err)
+
+	return token
+}
+
+func (f *fakeIDP) issuerURLTemplate() string {
+	return f.server.URL + "/org/%s"
+}
+
+func newFakeIDP(t *testing.T, organizationID string) *fakeIDP {
+	// Generate keypair for IDP
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	signer, err := jose.NewSigner(
+		jose.SigningKey{Algorithm: jose.RS256, Key: privateKey},
+		(&jose.SignerOptions{}).WithType("JWT"),
+	)
+	require.NoError(t, err)
+
+	providerMux := http.NewServeMux()
+	srv := httptest.NewServer(providerMux)
```

Comment:
```suggestion
	srv := httptest.NewServer(providerMux)
	t.Cleanup(src.Close)
```

Reply:
I was already doing this as part of the main test, but I'll move it into the helper for readabiltiy.

Reply:
a8c047cd6f